container/client.go

```diff
@@ -0,0 +1,211 @@
+
+	return dc, nil
+}
+
+func (dc *DockerClient) InspectContainer(id string) (*docker.Container, error) {
+	dc.inspectMutex.Lock()
+	defer dc.inspectMutex.Unlock()
+
+	// check if the inspect info is already stored in inspectCache
+	if info, haveInfo := dc.inspectCache[id]; haveInfo {
+		return info, nil
+	}
+
+	info, err := dc.cl.InspectContainer(id)
+	if err != nil {
+		return nil, err
+	}
+	dc.inspectCache[id] = info
+
+	return info, nil
+}
+
+func (dc *DockerClient) GetDockerParam(param string) (string, error) {
+	env, err := dc.cl.Info()
+	if err != nil {
+		log.WithFields(log.Fields{
+			"plugin": "docker collector",
+			"module": "client",
+			"block":  "GetCgroupDriver",
+		}).Error(err)
+		return "", err
+	}
+
+	if env.Exists(param) {
+		return env.Get(param), nil
+	}
+
+	log.WithFields(log.Fields{
+		"plugin": "docker collector",
+		"module": "client",
+		"block":  "GetCgroupDriver",
+	}).Errorf("%s setting not found", param)
+	return "", fmt.Errorf("%s not found", param)
+}
+
+// GetShortID returns short container ID (12 chars)
+func GetShortID(dockerID string) (string, error) {
+	if dockerID == "root" {
+		return dockerID, nil
+	}
+
+	if len(dockerID) < 12 {
+		return "", fmt.Errorf("Docker id %v is too short (the length of id should equal at least 12)", dockerID)
+	}
+
+	return dockerID[:12], nil
+}
+
+// ListContainersAsMap returns list of all available docker containers and base information about them (status, uptime, etc.)
+func (dc *DockerClient) ListContainersAsMap() (map[string]*ContainerData, error) {
+	containers := make(map[string]*ContainerData)
+
+	containerList, err := dc.cl.ListContainers(docker.ListContainersOptions{})
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, c := range containerList {
+		shortID, err := GetShortID(c.ID)
+		if err != nil {
+			return nil, err
+		}
+
+		spec := Specification{
+			Status:     c.Status,
+			Created:    time.Unix(c.Created, 0).Format("2006-01-02T15:04:05Z07:00"),
+			Image:      c.Image,
+			SizeRw:     c.SizeRw,
+			SizeRootFs: c.SizeRootFs,
+			Labels:     c.Labels,
+		}
+
+		containerData := ContainerData{
+			ID:            c.ID,
+			Specification: spec,
+			Stats:         NewStatistics(),
+		}
+
+		containers[shortID] = &containerData
+	}
+
+	containers["root"] = &ContainerData{ID: "/", Stats: NewStatistics()}
+
+	if len(containers) == 0 {
+		return nil, fmt.Errorf("No docker containers found %d", len(containers))
+	}
+
+	return containers, nil
+}
+
+// FindCgroupMountpoint returns cgroup mountpoint of a given subsystem
+func (dc *DockerClient) FindCgroupMountpoint(procfs string, subsystem string) (string, error) {
+	f, err := os.Open(filepath.Join(procfs, "self/mountinfo"))
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		txt := scanner.Text()
+		fields := strings.Split(txt, " ")
+		for _, opt := range strings.Split(fields[len(fields)-1], ",") {
+			if opt == subsystem {
+				return fields[4], nil
```

Comment:
It would require to `/proc/self/mountinfo` change its format. I think it's not needed